Elastic Average optimizer

jinxin0924 opened the pull request:
I have implemented a new distributed optimizer "ElasticAverageOptimizer". It is mentioned in the following issue:
#12472

EASGD is an async optimizer. During the training, Each worker will update the local variables and maintains its own local_step, which starts from 0 and is incremented by 1 after each update of local variables. Whenever the communication period divides the local step, the worker requests the current global center variables and then computed the elastic difference between global center variables and local variables. The elastic difference then be used to update both local variables and global variables.

**[Reference]**: https://arxiv.org/pdf/1412.6651.pdf

jinxin0924 committed:
easgd

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented:
@joshburkart could you take a look?

joshburkart commented:
Unfortunately I don't feel comfortable reviewing this... I don't know the paper referenced and I also am not super familiar with using TF in distributed contexts... Sorry!

yaroslavvb commented:
cc @ry who's reviewing related method in https://github.com/tensorflow/tensorflow/pull/11581

ry commented:
@jinxin0924 @yaroslavvb I'm happy to review it (which entails trying to repeat the results) but I won't have time for two weeks. Hit me up around Oct 2 if someone else hasn't yet.

jinxin0924 commented:
@ry Thanks a lot. Besides, I have also implemented model average, which may be better than #11581. Do you think it is appropriate to make a new pull request?

ry commented:
@jinxin0924 Let's do one at a time : ) make sure you have tests similar to what I proposed here https://github.com/ry/tensorflow/commit/764cd38b6407421e4fe23cf8cd9d2518ef043c8d
I wasn't able to get the original ModelAvg working due, I think, due to a bug demonstrated by that.

sb2nov commented:
@jinxin0924 can you take a look at the comment from @ry about adding a test.

jinxin0924 commented:
@sb2nov I think the test @ry mentioned  is about Model Average, other optimizer that I want to make pull request in the future, not this EASGD pr. :)

sb2nov commented:
@ry ping for review

DjangoPeng commented:
@jinxin0924 Could you attach some test results(like Execl tables, benchmark pictures, etc)? Moreover, would you completely implement the EASGD as the paper described?

@drpngx @joshburkart I'm familiar with EASGD, so I may help review the PR.

DjangoPeng commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,279 @@
+    with ops.device(self._worker_device):
+      local_var = getter(name, *args, **kwargs)
+
+    if kwargs['trainable']:
```
I think this may cause a KeyError when missing the key `trainable` in the kwargs. You'd better replace it with a more safe way.

jinxin0924 replied:
The customGetter is used in variable_scope, a context manager for defining ops that creates variables, so the key *trainable* should be in the kwargs. I think it is safe, but if you disagree, I can improve it.

DjangoPeng replied:
That sounds fine. I'm just worry about the wrong usage of `ElasticAverageCustomGetter`. When a TensorFlow freshman call `ElasticAverageCustomGetter` without variable_scope, it may make some trouble. We can't assume that the user input is always right.

alextp replied:
I prefer if you implement trainable as an explicit parameter to your custom_getter, like other custom_getters in tf are implemented.

DjangoPeng commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,279 @@
+    self._period = communication_period
+
+    if moving_rate is None:
+      self._moving_rate = 0.9 / communication_period / num_worker
```
Why 0.9 here? If it is reasonable, you'd better define a named constant instead of `0.9`.

DjangoPeng commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,279 @@
+    self._opt._prepare()
+
+  def compute_gradients(self, loss, *args, **kwargs):
+    # add rho*elastic_difference to loss to control the exploration
```
Add complete help document for `compute_gradients` function, including `Args`, `Return`, etc.

DjangoPeng commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,279 @@
+    return self._opt.compute_gradients(total_loss, *args, **kwargs)
+
+  def apply_gradients(self, grads_and_vars, global_step=None, name=None):
+    # update local variables
```
Add complete help document for `apply_gradients` function, including `Args`, `Return`, etc.

DjangoPeng commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,279 @@
+      local_center_vars = ops.get_collection_ref('local_center_variable')
+      local_center_vars_update = []
+      for lvar, var in zip(local_center_vars, global_center_vars):
+        # with ops.device(lvar.device):
```
What is the usage and purpose of the comment?

jinxin0924 committed:
Add complete help document

Add complete help document for apply_gradients ,compute_gradients. Define a named constant.

drpngx commented:
Jenkins, test this please.

drpngx commented:
@ry any thoughts on the newest code?

ry commented:
@drpngx I'm sorry, I've left Google, and cannot review this anymore. I apologize for not handing this off to someone else. 

@vrv might be able to suggest a new reviewer. (I have some code in my experimental directory called distbench to test distributed training, in case that's helpful.)

DjangoPeng commented:
@jinxin0924 Would you mind attaching some test results and test scripts for EASGD? I think that would help us judge and evaluate the PR.

@drpngx Personally speaking, the implementation and code are make sense. Hope get some advice from @vrv .

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+from tensorflow.python.framework import constant_op
+
+
+class ElasticAverageCustomGetter(object):
```
I think it might be better to expose a variable_scope object instead of a custom_getter directly, so users can say "with tf.contrib.op.elastic_average_variable_scope():"

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+  2. Generate global variables(global center variables)
+  3. Generate local variables(local center variables) which record the global
+    variables and place them at worker device
+  4. Remain global step variable as global variable
```
s/remin/rename? I'm not sure I understand the comment.

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+      local_var = getter(name, *args, **kwargs)
+
+    if kwargs['trainable']:
+      global_center_variable = variables.Variable(
```
Use variable_scope.variable instead of variables.Variable for future-proof support of resource variables.

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+    if kwargs['trainable']:
+      global_center_variable = variables.Variable(
+        name='%s/%s' %
+             ('global_center_variable',
```
Put this collection name in a named constant on top of the file. Maybe add underscores to it so it doesn't conflict with user-added collections?

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+          initial_value=local_var.initialized_value(),
+          trainable=False,
+          collections=[ops.GraphKeys.LOCAL_VARIABLES,
+                       'local_center_variable'])
```
Same thing for this collection name.

I wonder though if you're not better off using explicit data structures to communicate information from the getter to the optimizer. Maybe if you pass the getter as a constructor argument to the optimizer you can do that?

jinxin0924 replied:
I do not think passing the getter as a constructor to the optimizer will work. The local_var should be used to in forward computation and backward computation. But the global variables will be used in forward pass if the getter not used before building the model(getter returns the local_var).

alextp replied:
What I meant is that now you're using hidden global state in the optimizer to look up what is the global variable for a given local variable. Why not store a map in the custom_getter and make this map directly available to the optimizer by passing the custom_getter to the optimizer?

This way you do not need any new variable collections at all.

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+      variables.trainable_variables(),
+      ops.get_collection_ref('local_center_variable'))]
+
+    distance_loss = self._rho * \
```
I prefer to use parenthesis over \s to break lines.

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+                                       gate_gradients,aggregation_method,
+                                       colocate_gradients_with_ops,grad_loss)
+
+  def apply_gradients(self, grads_and_vars, global_step=None, name=None):
```
Do you really need to override apply_gradients? Can't you just override apply_dense/sparse instead?

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,336 @@
+    return sync_queue_op
+
+  def make_session_run_hook(self, is_chief, task_index):
+    """Creates a hook to handle SyncReplicasHook ops such as initialization."""
```
This doesn't look like it's doing anything sync-replicas-related. Stale comment?

jinxin0924 committed:
Update elastic_average_optimizer.py

jinxin0924 committed:
Update elastic_averaging_optimizer_test.py

jinxin0924 committed:
Update elastic_average_optimizer.py

jinxin0924 committed:
Update elastic_average_optimizer.py

jinxin0924 commented:
@alextp, thanks for review.
"I prefer if you implement trainable as an explicit parameter to your custom_getter, like other custom_getters in tf are implemented." - Could you please give me some examples?

yaroslavvb commented:
@jinxin0924 example https://github.com/tensorflow/tensorflow/blob/4e69e02241067129379f73dd4fefe57f0a12bdc9/tensorflow/python/training/moving_averages.py#L130

alextp commented on tensorflow/contrib/opt/python/training/elastic_average_optimizer.py:
```diff
@@ -0,0 +1,337 @@
+    # update global variables.
+    def _Update_global_variables():
+      local_vars = variables.trainable_variables()
+      global_center_vars = ops.get_collection_ref(GLOBAL_CENTER_VARIABLE)
```
This is updating more variables than those specified in grads_and_vars. This is not ideal.

jinxin0924 replied:
In EASGD, the vars in grads_and_vars should be updated at each step, and the global variables will be updated when the step is divided by communication_period. I think these variables update should be placed in apply_gradients function. Should I add local_vars and global_center_vars to the apply_gradients function as arguments?

alextp commented:
I agree these updates should be placed in the apply_gradients function, but
I think they can be handled independently in the apply_dense / apply_sparse
methods.

On Sun, Nov 12, 2017 at 5:19 PM, JxKing <notifications@github.com> wrote:

> *@jinxin0924* commented on this pull request.
> ------------------------------
>
> In tensorflow/contrib/opt/python/training/elastic_average_optimizer.py
> <https://github.com/tensorflow/tensorflow/pull/13012#discussion_r150432917>
> :
>
> > +      An `Operation` that applies the specified gradients. If `global_step`
> +      was not None, that operation also increments `global_step`.
> +
> +    Raises:
> +      TypeError: If `grads_and_vars` is malformed.
> +      ValueError: If none of the variables have gradients.
> +    """
> +    apply_updates = self._opt.apply_gradients(grads_and_vars)
> +    with ops.control_dependencies([apply_updates]):
> +      local_update = state_ops.assign_add(
> +        self._local_step, 1, name='local_step_update').op
> +
> +    # update global variables.
> +    def _Update_global_variables():
> +      local_vars = variables.trainable_variables()
> +      global_center_vars = ops.get_collection_ref(GLOBAL_CENTER_VARIABLE)
>
> In EASGD, the vars in grads_and_vars should be updated at each step, and
> the global variables will be updated when the step is divided by
> communication_period. I think these variables update should be placed in
> apply_gradients function. Should I add local_vars and global_center_vars to
> the apply_gradients function as arguments?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/13012#discussion_r150432917>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxVXHHahWWs22bf77OkU790NBNkJbks5s15kxgaJpZM4PV-8Y>
> .
>



-- 
 - Alex

jinxin0924 committed:
Update elastic_average_optimizer.py

store a map in the custom_getter and  pass it to the optimizer

jinxin0924 committed:
Update elastic_averaging_optimizer_test.py

jinxin0924 commented:
@alextp, I have updated the custome_getter. Now the customer_getter has two maps and is directly available to the optimizer.
I do not understand why the updates of global/local center variables will be better in apply_dense / apply_sparse, which may needs much more codes to implement.

DjangoPeng commented:
@jinxin0924 I think you have to rebase your dev branch to solve the conflict.

jinxin0924 commented:
@DjangoPeng , Sure, thanks for reminding.

jinxin0924 committed:
resolve confilicts

jinxin0924 committed:
Merge branch 'master' into easgd

alextp reviewed (approved):
Ok, I think this looks good now. Thanks for the changes!

alextp commented:
Jenkins, test this please.

DjangoPeng reviewed (approved):
@jinxin0924 LGTM. Approve this pull request.

jinxin0924 committed:
Update BUILD

jinxin0924 committed:
Update BUILD

jinxin0924 committed:
Update BUILD

jinxin0924 committed:
Update BUILD

jinxin0924 committed:
Update elastic_average_optimizer.py

jinxin0924 committed:
Update elastic_average_optimizer.py

jinxin0924 committed:
Merge branch 'master' into easgd

jinxin0924 commented:
@alextp, could you help me with the failed checks? I did not understand what happened.

alextp commented:
Jenkins, test this please.

alextp commented:
It claimed an infrastructure error so let's try again.

alextp commented:
Traceback (most recent call last):
  File "/Volumes/BuildData/tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/darwin_x86_64-opt/bin/tensorflow/tools/test/check_futures_test.runfiles/org_tensorflow/tensorflow/tools/test/check_futures_test.py", line 107, in <module>
    main()
  File "/Volumes/BuildData/tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/darwin_x86_64-opt/bin/tensorflow/tools/test/check_futures_test.runfiles/org_tensorflow/tensorflow/tools/test/check_futures_test.py", line 103, in main
    raise AssertionError('Error in %s: %s' % (short_path, str(e)))
AssertionError: Error in contrib/opt/python/training/elastic_averaging_optimizer_test.py: Missing futures: division absolute_import print_function

To fix this you need to "from future import .."; copy all the imports in the other tensorflow python files.

jinxin0924 committed:
Update elastic_averaging_optimizer_test.py

jinxin0924 commented:
@alextp, thanks so much.

caisq commented:
@tensorflow-jenkins test this please

caisq commented:
@jinxin0924 It looks like a test file is missing from this PR:
https://ci.tensorflow.org/job/tensorflow-pull-requests-gpu/7893/consoleFull

> ERROR: missing input file '//tensorflow/contrib/opt:python/training/elastic_average_optimizer_test.py'.
> ERROR: /workspace/tensorflow/contrib/opt/BUILD:177:1: //tensorflow/contrib/opt:elastic_average_optimizer_test: missing input file '//tensorflow/contrib/opt:python/training/elastic_average_optimizer_test.py'.
> INFO: Building complete.
> ERROR: /workspace/tensorflow/contrib/opt/BUILD:177:1 1 input file(s) do not exist.

jinxin0924 committed:
Rename elastic_averaging_optimizer_test.py to elastic_average_optimizer_test.py

jinxin0924 commented:
@caisq , It was a stupid mistake... I have renamed the test file.

caisq commented:
@tensorflow-jenkins test this please

caisq commented:
@jinxin0924 There are test failures. For example: https://ci.tensorflow.org/job/tensorflow-pull-requests-cpu-python3/7533/consoleFull

Please take a look and see if you can fix them.

jinxin0924 committed:
Update elastic_average_optimizer_test.py

zip is different in py2 and py3

jinxin0924 commented:
@caisq It was the zip difference in py2 and py3 which caused failures.  I have modified it and tested both in py2 and py3.

drpngx commented:
Jenkins, test this please

On Sun, Dec 10, 2017, 9:45 PM JxKing <notifications@github.com> wrote:

> @caisq <https://github.com/caisq> It was the zip difference in py2 and
> py3 which caused failures. I have modified it and tested both in py2 and
> py3.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/13012#issuecomment-350627738>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbRCJ22v9YlcYqPCB4bua9l7bXMrdks5s_MGDgaJpZM4PV-8Y>
> .
>

caisq commented:
Thanks, @jinxin0924 !

DjangoPeng commented:
@jinxin0924 Congratulations! 🎉

jinxin0924 commented:
@caisq @DjangoPeng Thansk!
